archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.fintype.basic
+import data.fintype.card
+import data.equiv.basic
+import group_theory.group_action
+
+import tactic.field_simp
+import tactic.ring
+
+/-!
+# Freek № 88 : Derangements Formula
+This file proves Theorem 88 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/),
+which gives an explicit formula for the number of derangements (permutations without fixed
+points) on `n` elements.
+
+The core of the proof is an equivalence between "derangements on `n+2` elements", and
+an expression involving derangements on `n` and `n+1` elements. This gives a recurrence
+relation `d(n+2) = (n+1) * (d(n) + d(n+1))`, from which a sum-based formula can be derived.
+-/
+
+open equiv fintype
+open_locale big_operators
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
```

Comment:
First comment --- absolutely everything in this PR belongs in mathlib proper. No need to add anything to `archive/100-theorems-list/` except a pointer to the result.

This will require moving things to a bunch of different files!

Reply:
would it go somewhere like `combinatorics/derangements.lean` and `data/equiv/<something>.lean`? (depending on if it's about perms or derangements specifically)

Reply:
I imagine that the default location for things moved from this should be to a new file `data/equiv/perm/derangements.lean`

Reply:
@HenrySwanson, when a conversation has been dealt with, it's easiest if the author takes the initiative to click "Resolve conversation". I think this one is good now?